File: LLD/Design-Patterns/Prototype/code/prototype.py
"""
Prototype Pattern — cloning a game Enemy, demonstrating deep copy of a nested field.
Run: python prototype.py
"""

import copy


class Stats:
    def __init__(self, health, damage):
        self.health = health
        self.damage = damage


class Enemy:
    def __init__(self, enemy_type, stats):
        self.type = enemy_type
        self.stats = stats

    def clone(self):
        # copy.deepcopy recursively clones nested objects (like `stats`)
        # instead of sharing references with the original.
        return copy.deepcopy(self)

    def __repr__(self):
        return f"{self.type}{{health={self.stats.health}, damage={self.stats.damage}}}"


if __name__ == "__main__":
    orc_prototype = Enemy("Orc", Stats(100, 15))

    orc1 = orc_prototype.clone()
    orc2 = orc_prototype.clone()

    # Mutate orc1's stats — orc2 and the prototype must stay unaffected.
    orc1.stats.health = 40

    print("Prototype:", orc_prototype)
    print("orc1 (damaged):", orc1)
    print("orc2 (untouched):", orc2)

File: v1/LLD/Design-Patterns/Prototype/code/Prototype.cpp
// Prototype Pattern — cloning a game Enemy, demonstrating deep copy of a nested field.
// Compile: g++ -std=c++14 Prototype.cpp -o prototype
// Run:     ./prototype

#include <iostream>
#include <string>
#include <memory>

class Stats {
public:
    int health;
    int damage;

    Stats(int health, int damage) : health(health), damage(damage) {}

    std::unique_ptr<Stats> deepCopy() const {
        return std::make_unique<Stats>(health, damage);
    }
};

class Enemy {
public:
    std::string type;
    std::unique_ptr<Stats> stats;

    Enemy(std::string type, std::unique_ptr<Stats> stats)
        : type(std::move(type)), stats(std::move(stats)) {}

    // Deep copy: stats gets its own independent copy, not a shared pointer.
    std::unique_ptr<Enemy> clone() const {
        return std::make_unique<Enemy>(type, stats->deepCopy());
    }

    void print() const {
        std::cout << type << "{health=" << stats->health
                   << ", damage=" << stats->damage << "}" << std::endl;
    }
};

int main() {
    Enemy orcPrototype("Orc", std::make_unique<Stats>(100, 15));

    auto orc1 = orcPrototype.clone();
    auto orc2 = orcPrototype.clone();

    // Mutate orc1's stats — orc2 and the prototype must stay unaffected.
    orc1->stats->health = 40;

    std::cout << "Prototype: ";
    orcPrototype.print();

    std::cout << "orc1 (damaged): ";
    orc1->print();

    std::cout << "orc2 (untouched): ";
    orc2->print();

    return 0;
}
